Develop/holland01

holland01 opened the pull request:
Just support for MSVC. Note that ssize_t is a posix extension which Windows doesn't appear to support still.

holland01 committed:
support for compilers which don't support dynamic array initialization syntax

holland01 committed:
windows compat

holland01 committed:
tested extension with msvc 2017; it doesn't work either

holland01 commented on base58.c:
```diff
@@ -37,39 +49,69 @@ typedef uint32_t b58_almostmaxint_t;
 static const b58_almostmaxint_t b58_almostmaxint_mask = ((((b58_maxint_t)1) << b58_almostmaxint_bits) - 1);
 
+//	MSVC 2017 doesn't support the GCC (and probably clang) extension
+//	for dynamic arrays in C
```
>Variable length arrays are a standard C99 feature, not an extension. 

That may be, but MSVC still doesn't support them. _alloca or the MSVC _malloca  can be used if that's preferred - these perform stack allocation (though _malloca will allocate for sufficiently large sizes).

luke-jr replied:
AFAIK, MSVC is a C++ compiler, not a C compiler.

I suppose it can't hurt to use _malloca or such, based on a failed configure check for standards compatibility...

holland01 replied:
A few years back they added C99 support. It's still subpar in comparison to MinGW/GCC, but it's better than it used to be - enough to where people who like C end up using it over C++.

luke-jr replied:
Wrote a simple alternative in #9; can you check if it works?

holland01 commented on base58.c:
```diff
@@ -5,19 +5,31 @@
  */
 
-#ifndef WIN32
+#include "libbase58.h"
```
stdint.h is in libbase58.h - see [here](https://github.com/holland01/libbase58/blob/develop/holland01/libbase58.h#L6).

I also didn't see any reason include 4 header files before the header file, considering that most of those aren't necessary for the function declarations found in it.

holland01 commented on base58.c:
```diff
@@ -5,19 +5,31 @@
 
 #include <stdbool.h>
 #include <stddef.h>
-#include <stdint.h>
```
Yes, and it's [declared in the header file](stdint.h is in libbase58.h - see [here](https://github.com/holland01/libbase58/blob/develop/holland01/libbase58.h#L6).)

luke-jr replied:
It should be included in files where it is used directly.

holland01 commented on base58.c:
```diff
@@ -5,19 +5,31 @@
+#ifdef _WIN64
+typedef int64_t ssize_t;
+#else
+typedef int32_t ssize_t;
```
I'm honestly wondering why `ssize_t` existed there in the first place, then?

luke-jr replied:
Convenience; `j` can get to `-1` with the current code. So we need to handle it differently to avoid that.

luke-jr replied:
Moved this issue to #8; please review

holland01 commented on base58.c:
```diff
@@ -5,19 +5,31 @@
 #include <arpa/inet.h>
 #else
 #include <winsock2.h>
+#include <malloc.h>
```
I'll replace with an appropriate `_MSC_VER` check.

holland01 commented on base58.c:
```diff
@@ -37,39 +49,69 @@ typedef uint32_t b58_almostmaxint_t;
 	b58_almostmaxint_t zeromask = bytesleft ? (b58_almostmaxint_mask << (bytesleft * 8)) : 0;
 	unsigned zerocount = 0;
-	
+
```
Apologies, I'll be sure to omit any of those in the future.

luke-jr committed:
tests: Check encoding with highest bit set

Detects j dropping below 0 going unchecked.
This bug was originally fixed in 2c6b7916b604b946904428466ff0271a35535297, but due to ssize_t being a POSIX extension, we are going to check it manually.

luke-jr committed:
Avoid ssize_t type, since it is POSIX-specific

Only j could become negative, so we simply check before it would

holland01 committed:
merging in changes from luke-jr wrt ssize_t fix (avoid_ssize_t)

holland01 committed:
requested modifications for MSVC fix

holland01 commented:
@luke-jr Ok, changes have been added. Please review

holland01 commented on base58.c:
```diff
@@ -41,26 +43,25 @@ typedef uint32_t b58_almostmaxint_t;
+#ifdef _MSC_VER
 #define b58_log_err(msg, ...) printf("[" __FUNCTION__ " - ERROR]: " msg, __VA_ARGS__)
 
-#define b58_alloc_mem(type, name, count)					\
```
I'm honestly not sure what's up with the indentation of the escape slashes here. With invisibles in Atom turned on everything looks as if it's tabbed appropriately, and this is exactly how it looked before the push:

![badformat](https://user-images.githubusercontent.com/911971/40892206-7bece1a8-6748-11e8-99b4-ca55f2bb4476.PNG)

jasny commented:
@luke-jr :ping_pong:

What is needed to merge this PR and fix the Windows build?

luke-jr commented:
At the very least, a clean and reviewable branch...
